Add approvers, approval timestamp and URS text to verification report

sebastianden opened the pull request:
Main:
- Changed `render_requirements.py` to add a second part in the Requirements section of the verification report to include:
    - Approvers of the PR the feature file was changed
    - Approval timestamp of PR where feature was changed
    - Current text of the feature file
- This change improves audit readiness drastically
- Added option of providing a roles file (.yml) in to allow for a mapping of novo initials to actual names and roles in the project  

Outcome:
<img width="1698" alt="new-section" src="https://github.com/user-attachments/assets/15bdebea-b254-4c39-a631-e65b2eb0ff34" />

Without providing optional roles.yml:
<img width="702" alt="no-roles-file" src="https://github.com/user-attachments/assets/72a3039e-e1d0-4f75-a5ec-20a4f8e2b8e6" />

Also works if no one approved PR (though I'm wondering if it even should):
<img width="623" alt="no-approver" src="https://github.com/user-attachments/assets/93a2819c-05e4-4189-b73c-1bdca0723982" />

lasselundstenjensen commented on templates/scripts/report/render_requirements.py:
```diff
@@ -1,59 +1,181 @@
+        # If the commit id is in the PR commits, return the PR id and completion date
+        if any(commit["commitId"] == commit_id for commit in pr_commits):
+            # Remove milliseconds
+            completion_date = pr["closedDate"].split(".")[0] + "Z"
+            return (pr["pullRequestId"], completion_date)
+    return None
+
+
+def get_approvers(pr_id: str, pull_requests: list[dict]) -> list[str]:
+    """Get the approvers for a given pull request."""
+
+    for pr in pull_requests:
+
+        if pr["pullRequestId"] == pr_id:
+            reviewers = pr["reviewers"]
+            approvers = [
+                reviewer["uniqueName"]
+                for reviewer in reviewers
+                if reviewer["vote"] == 10
+            ]
+            return approvers
+    else:
+        return []
+
+
+def get_approver_roles(approvers: list[str], roles: dict) -> list[str]:
+    """Get the roles of the approvers by matching initials in the roles.yml
+    file.
+    """
+    approvers_with_roles = []
+    for approver in approvers:
+        # Try to get the roles of the approvers
+        if (initials := approver.split("@")[0].lower()) in roles:
+            approvers_with_roles.append(
+                f"{roles[initials]['name']}, {roles[initials]['role']}"
+            )
+        else:
+            approvers_with_roles.append(approver)
+    return approvers_with_roles
+
```
Please extract this into a separate file for better readability. We have similar files for talking to the API (e.g. [`get_pull_request_id.py`](https://github.com/nn-dma/qms-pipeline-templates/blob/main/templates/scripts/report/get_pull_request_id.py)). Name is legacy, but many other calls reside in separate files to at least separate concerns.

The `generate_verification_report.py` file is supposed to do orchestration only.

sebastianden replied:
Done

lasselundstenjensen commented on templates/scripts/report/render_requirements.py:
```diff
@@ -108,38 +261,94 @@ def main(argv):
+
+        # Base URL for Azure DevOps REST API
+        base_url = f"https://dev.azure.com/{organization}/{project}/_apis/git/repositories/{repository}"
+
+        pull_requests = get_pull_requests(base_url, auth_method, access_token, branch)
+
+        for file in files:
+
+            with open(file, mode="r", encoding="utf-8") as file_reader:
+                lines_raw = file_reader.read()
+                lines = lines_raw.split("\n")
+                features = extract_features(lines)
+
+            last_modified_commit_hash = extract_last_modified_commit_hash(file, branch)
+
+            pr_id, completion_date = match_commit_with_pr(
+                base_url,
+                auth_method,
+                access_token,
+                pull_requests,
+                last_modified_commit_hash,
+            )
+            approvers = get_approvers(pr_id, pull_requests)
+
+            if role_file:
+                approvers = get_approver_roles(approvers, roles)
+
+            for feature in features:
+                print(
+                    f"""    <h4>{render_tags(feature.tags)}</h4>
+    <pre>{lines_raw}</pre>
+    <p>Approved at: {completion_date}</p>
+    <p>Approvers:</p>
+    <ul>"""
+                )
+                for approver in approvers:
+                    print(f"        <li>{approver}</li>")
+                print("""    </ul>""")
+        print("</figure>")
+
```
Please extract this into a separate rendering file, e.g. like [`render_design_specifications`](https://github.com/nn-dma/qms-pipeline-templates/blob/main/templates/scripts/report/render_design_specifications.py).

This makes it testable separately as a standalone script, i.e. you can generate the HTML from input that fulfils the input "schema".

You would then call it in this kind of fashion to inject the rendered HTML into the verification report template. I would suggest maybe even going as far as creating a new PLACEHOLDER (instead of here `<var>LIST_OF_DESIGN_SPECIFICATIONS</var>`) in the template for this to make its existance more obvious:

```bash
echo -e "================> ${PURPLE}Extracting and rendering design specifications${NC}"
python3 ../${{ parameters.render_design_specifications_py_location }} -folder ${{ parameters.system_design_path }} -branch origin/release/$(Build.SourceBranchName) -organization novonordiskit -project '$(System.TeamProject)' -repository $(Build.Repository.Name) > listOfDesignSpecifications.html
python3 ../${{ parameters.render_replace_py_location }} -render ./listOfDesignSpecifications.html -template ../${{ parameters.verification_report_template_location }} -placeholder "<var>LIST_OF_DESIGN_SPECIFICATIONS</var>"
```

lasselundstenjensen commented on templates/scripts/report/render_requirements.py:
```diff
@@ -1,59 +1,181 @@
+import sys
+
+import requests
+import yaml
```
Move to extracted file as mentioned in below comment.

sebastianden replied:
Done

lasselundstenjensen reviewed (changes requested):
Agree that it improves inspection readiness.

My review is focusing on this particular implementation, which is tied to ADO only. But we also have to keep in mind that this will create a new flavor of the verification report that QCs will see. And as such this functionality would need to be ported to the Github version as well. The Github implementation is a lot more detached from the pipeline workflow as a whole, but the API calls are also different and we may not be able to get the same outputs which you are putting in the ADO version (here) of the report.

Currently, the two reports are identical in shape and content regardless of being generated from Github or ADO. With this change, they will be different.

Part of this change and review will also be asking you to look into whether the same information can be gathered from the Github APIs as to consolidate the report generation there.

fadur commented:
@sebastianden I think you should use this as your own fork.
I dont want to ask you to impliment the same for github as well. And I agree with @lasselundstenjensen, we really dont want to have two versions of this.

lasselundstenjensen commented:
@sebastianden 
If we change the report in one place (ADO), we should change it for Github as well so QCs across Novo do not encounter different flavors of the report. This repo (`qms-pipeline-templates`) only serves ADO. Alternatively, as @fadur also points out, **you _could_ opt to use this specifically for your pipelines only (e.g. via a fork).**

The way the report is generated in Github is via a Dagger workflow, and the structure is a little different than for ADO—but not much.

API calls will be different though and the code to parse them and what data is available is likely also going to be different.

lasselundstenjensen commented:
@sebastianden 
**Another question:**

Have you tested what the output and HTML injected into the verification report will be in case of any error (fatal and other types) in calling the ADO API or processing e.g. a HTTP response that does not follow the pattern you expect?

If no data can be found, the HTML should be blank (this is how the rest of the injection mechanisms work in the verification report generation.

sebastianden commented:
@fadur @lasselundstenjensen I'd like to avoid using my own fork as it might diverge from this central implementation in the future. I agree that this should also be in the Github implementation and believe that it adds a lot of value for everyone using the QMS pipeline. I cannot promise to implement it singlehandedly like I did here, but I can offer you to investigate the GH API and sit down with potentially @alegarau to implement it. (I'm not familiar with dagger).

@lasselundstenjensen I have tested multiple possible failure scenarios, such as no approvers at all (which shouldn't be but is technically possible) and of course no roles file provided in which case it defaults to the normal email addresses with initials. See screenshots for both. I only make two API calls: One to get all pull requests into the current branch (which there must be at least one, else there would be no release) and then to get the commit hashes in the PRs (again the files last changed commit hash must appear in one of them else the file wouldn't be here). 

PS: I addressed the comments above regarding code in this repository. I'm more than happy to update the docs if we can agree to merge this.
